### PR TITLE
Remove SIMD Feature

### LINKED ISSUE #46: Fail to compile with unrecognized platform-specific intrinsic function
*Note*: migrated from original JIRA: https://issues.apache.org/jira/browse/ARROW-5613

I'm testing a project which depends on the Arrow crate. It failed with the following error:
{code}
error[E0441]: unrecognized platform-specific intrinsic function: `simd_bitmask`
   --> /Users/sunchao/.cargo/registry/src/github.com-1ecc6299db9ec823/packed_simd-0.3.3/src/codegen/llvm.rs:100:5
    |
100 |     crate fn simd_bitmask<T, U>(value: T) -> U;
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to previous error

For more information about this error, try `rustc --explain E0441`.
error: Could not compile `packed_simd`.
{code}

### LINKED ISSUE #54: packed_simd requires nightly
*Note*: migrated from original JIRA: https://issues.apache.org/jira/browse/ARROW-6718

See [https://github.com/rust-lang/rfcs/pull/2366] for more info on stabilization of this crate.

 
{code:java}
error[E0554]: `#![feature]` may not be used on the stable release channel
   --> /home/andy/.cargo/registry/src/github.com-1ecc6299db9ec823/packed_simd-0.3.3/src/lib.rs:202:1
    |
202 | / #![feature(
203 | |     repr_simd,
204 | |     const_fn,
205 | |     platform_intrinsics,
...   |
215 | |     custom_inner_attributes
216 | | )]
    | |__^
 {code}

### LINKED ISSUE #2856: Miri failures in simd aggregation using latest nightly
**Describe the bug**

I think this is not a new bug, but rather something that miri is now able to detect.

From my understanding this is a bug in packed_simd2:

https://github.com/rust-lang/packed_simd/blob/e4ec7ce86ba5e6479409c91b9a9a6af25536b047/src/api/slice/write_to_slice.rs#L63

```rust
pub unsafe fn write_to_slice_unaligned_unchecked(self, slice: &mut [$elem_ty]) {
    debug_assert!(slice.len() >= $elem_count);
    let target_ptr = slice.get_unchecked_mut(0) as *mut $elem_ty as *mut u8; // <-- here
    let self_ptr = &self as *const Self as *const u8;
    crate::ptr::copy_nonoverlapping(self_ptr, target_ptr, crate::mem::size_of::<Self>());
}
```

That line takes a mutable reference to the first element of the slice and then uses that as a pointer to write multiple elements. This would mean it is not actually writing out of bounds of the underlying allocation, but would be violating the stacked borrows model.

Since it is unlikely that packed_simd2 gets bugfix releases, this might be a reason to migrate to portable_simd instead.

**To Reproduce**

```
$ cargo --version
cargo 1.66.0-nightly (0b84a35c2 2022-10-03)
```

```
$ cargo miri test --features simd -- test_primitive_array_sum
test compute::kernels::aggregate::tests::test_primitive_array_sum ... error: Undefined Behavior: attempting a write access using <5683393> at alloc2379099[0x4], but that tag does not exist in the borrow stack for this location
   --> /home/i526205/.cargo/registry/src/github.com-1ecc6299db9ec823/packed_simd_2-0.3.8/src/v512.rs:50:1
    |
50  | / impl_i!([i32; 16]: i32x16, m32x16 | i32, u16 | test_v512 |
51  | |         x0, x1, x2, x3, x4, x5, x6, x7, x8, x9, x10, x11, x12, x13, x14, x15 |
52  | |         From: i8x16, u8x16, i16x16, u16x16 |
53  | |         /// A 512-bit vector with 16 `i32` lanes.
54  | | );
    | | ^
    | | |
    | |_attempting a write access using <5683393> at alloc2379099[0x4], but that tag does not exist in the borrow stack for this location
    |   this error occurs as part of an access at alloc2379099[0x0..0x40]
    |
    = help: this indicates a potential bug in the program: it performed an invalid operation, but the Stacked Borrows rules it violated are still experimental
    = help: see https://github.com/rust-lang/unsafe-code-guidelines/blob/master/wip/stacked-borrows.md for further information
help: <5683393> was created by a SharedReadWrite retag at offsets [0x0..0x4]
   --> arrow/src/datatypes/numeric.rs:323:26
    |
323 |                 unsafe { simd_result.write_to_slice_unaligned_unchecked(slice) };
    |                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
...
342 | make_numeric_type!(Int32Type, i32, i32x16, m32x16);
    | -------------------------------------------------- in this macro invocation
    = note: BACKTRACE:
    = note: inside `packed_simd_2::v512::<impl packed_simd_2::Simd<[i32; 16]>>::write_to_slice_unaligned_unchecked` at /home/i526205/.cargo/registry/src/github.com-1ecc6299db9ec823/packed_simd_2-0.3.8/src/api/slice/write_to_slice.rs:65:17
note: inside `<datatypes::types::Int32Type as datatypes::numeric::ArrowNumericType>::write` at arrow/src/datatypes/numeric.rs:323:26
   --> arrow/src/datatypes/numeric.rs:323:26
    |
323 |                 unsafe { simd_result.write_to_slice_unaligned_unchecked(slice) };
    |                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
...
342 | make_numeric_type!(Int32Type, i32, i32x16, m32x16);
    | -------------------------------------------------- in this macro invocation
note: inside `<compute::kernels::aggregate::simd::SumAggregate<datatypes::types::Int32Type> as compute::kernels::aggregate::simd::SimdAggregate<datatypes::types::Int32Type>>::reduce` at arrow/src/compute/kernels/aggregate.rs:390:13
   --> arrow/src/compute/kernels/aggregate.rs:390:13
    |
390 |             T::write(simd_accumulator, slice);
    |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: inside `compute::kernels::aggregate::simd::simd_aggregation::<datatypes::types::Int32Type, compute::kernels::aggregate::simd::SumAggregate<datatypes::types::Int32Type>>` at arrow/src/compute/kernels/aggregate.rs:641:9
   --> arrow/src/compute/kernels/aggregate.rs:641:9
    |
641 |         A::reduce(chunk_acc, rem_acc)
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: inside `compute::kernels::aggregate::sum::<datatypes::types::Int32Type>` at arrow/src/compute/kernels/aggregate.rs:655:5
   --> arrow/src/compute/kernels/aggregate.rs:655:5
    |
655 |     simd::simd_aggregation::<T, SumAggregate<T>>(&array)
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: inside `compute::kernels::aggregate::tests::test_primitive_array_sum` at arrow/src/compute/kernels/aggregate.rs:692:24
   --> arrow/src/compute/kernels/aggregate.rs:692:24
    |
692 |         assert_eq!(15, sum(&a).unwrap());
    |                        ^^^^^^^
note: inside closure at arrow/src/compute/kernels/aggregate.rs:690:5
   --> arrow/src/compute/kernels/aggregate.rs:690:5
    |
689 |       #[test]
    |       ------- in this procedural macro expansion
690 | /     fn test_primitive_array_sum() {
691 | |         let a = Int32Array::from(vec![1, 2, 3, 4, 5]);
692 | |         assert_eq!(15, sum(&a).unwrap());
693 | |     }
    | |_____^
```

**Expected behavior**

Tests should run without miri failures

**Additional context**

Migration to portable_simd was previously discussed in #1492

### LINKED ISSUE #5185: Remove nightly-only simd feature and related code in ArrowNumericType
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

With the merge of #5100 all users of the feature-gated methods in `ArrowNumericType` should be gone and those functions can be removed.

**Describe the solution you'd like**

 - Remove simd related functions from `ArrowNumericType` trait
 - Remove `bit_util::bitwise_bin_op_simd` which seems to be unused
 - Remove `simd` feature and `packed_simd` dependency

**Describe alternatives you've considered**

The `ArrowNumericType` seems a little bit redundant with `ArrowNativeTypeOp`, but removing it completely would involve a lot of little api changes, so it's probably better to keep it as an empty marker trait.

**Additional context**

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #46
Closes #54
Closes #2856
Closes #5185

# Rationale for this change
 

#5100 removed the last major use of SIMD, so we can now remove this feature. I debated deprecating things first, but this got complicated as you can't really deprecate a trait, so I figured I would at least propose just ripping the band-aid off in one.

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: arrow-ord/src/comparison.rs [hunk 1]
```diff
@@ -610,7 +552,6 @@ pub fn gt_eq_utf8_scalar<OffsetSize: OffsetSizeTrait>(
 /// Perform `left == right` operation on an array and a numeric scalar
 /// value. Supports PrimitiveArrays, and DictionaryArrays that have primitive values.
 ///
-/// If `simd` feature flag is not enabled:
```
Comment (tustvold) [inline arrow-ord/src/comparison.rs:613]: This hasn't actually been true since #4701

File: arrow-buffer/src/util/bit_util.rs [hunk 2]
```diff
@@ -104,31 +101,13 @@ pub fn ceil(value: usize, divisor: usize) -> usize {
...
 }
 
-/// Performs SIMD bitwise binary operations.
-///
-/// # Safety
-///
-/// Note that each slice should be 64 bytes and it is the callers responsibility to ensure
-/// that this is the case.  If passed slices larger than 64 bytes the operation will only
-/// be performed on the first 64 bytes.  Slices less than 64 bytes will panic.
-#[cfg(feature = "simd")]
```
Comment (tustvold) [inline arrow-buffer/src/util/bit_util.rs:114]: Interestingly arrow-buffer doesn't have a SIMD feature, so this has been silently not present by accident for a while

**Comment by jhorstmann** [comment]
Ok, that then also fixes #5185 which I just created :)

**Review by jackwener — APPROVED** [review]
Great job. Thanks @tustvold 
Appreciate #5100 of @jhorstmann

**Pull request merged**